File: bindings/typescript/src/tools.js
import { RustCrateBridge } from "./bridge.js";

export class SearchTools {
  static search(provider, query, config = {}) {
    return RustCrateBridge.invoke("or-tools-search", "search", {
      provider,
      query,
      config,
    });
  }
}

export class WebTools {
  static fetch(provider, request, config = {}) {
    return RustCrateBridge.invoke("or-tools-web", "fetch", {
      provider,
      request,
      config,
    });
  }

  static scrape(provider, url, config = {}) {
    return RustCrateBridge.invoke("or-tools-web", "scrape", {
      provider,
      url,
      config,
    });
  }
}

export class VectorTools {
  static ensureCollection(provider, data, config = {}) {
    return RustCrateBridge.invoke("or-tools-vector", "ensure_collection", {
      provider,
      data,
      config,
    });
  }

  static upsert(provider, data, config = {}) {
    return RustCrateBridge.invoke("or-tools-vector", "upsert", {
      provider,
      data,
      config,
    });
  }

  static delete(provider, data, config = {}) {
    return RustCrateBridge.invoke("or-tools-vector", "delete", {
      provider,
      data,
      config,
    });
  }

  static query(provider, data, config = {}) {
    return RustCrateBridge.invoke("or-tools-vector", "query", {
      provider,
      data,
      config,
    });
  }
}

export class LoaderTools {
  static load(request) {
    return RustCrateBridge.invoke("or-tools-loaders", "load", { request });
  }
}

export class ExecTools {
  static execute(request, providers = ["python", "shell"], config = {}) {
    return RustCrateBridge.invoke("or-tools-exec", "execute", {
      request,
      providers,
      config,
    });
  }
}

export class FileTools {
  static read(path, provider = "local", config = {}) {
    return RustCrateBridge.invoke("or-tools-file", "read", {
      provider,
      path,
      config,
    });
  }

  static write(path, content, provider = "local", config = {}) {
    return RustCrateBridge.invoke("or-tools-file", "write", {
      provider,
      path,
      content,
      config,
    });
  }

  static list(path, provider = "local", config = {}) {
    return RustCrateBridge.invoke("or-tools-file", "list", {
      provider,
      path,
      config,
    });
  }

  static delete(path, provider = "local", config = {}) {
    return RustCrateBridge.invoke("or-tools-file", "delete", {
      provider,
      path,
      config,
    });
  }

  static fetch(provider, query, config = {}) {
    return RustCrateBridge.invoke("or-tools-file", "fetch", {
      provider,
      query,
      config,
    });
  }
}

export class CommsTools {
  static send(provider, to, body, from = null, config = {}) {
    return RustCrateBridge.invoke("or-tools-comms", "send", {
      provider,
      to,
      body,
      from,
      config,
    });
  }
}

export class ProductivityTools {
  static listEmails(provider, query = {}, config = {}) {
    return RustCrateBridge.invoke("or-tools-productivity", "list_emails", {
      provider,
      query,
      config,
    });
  }

  static sendEmail(provider, item, config = {}) {
    return RustCrateBridge.invoke("or-tools-productivity", "send_email", {
      provider,
      item,
      config,
    });
  }

  static listEvents(provider, query = {}, config = {}) {
    return RustCrateBridge.invoke("or-tools-productivity", "list_events", {
      provider,
      query,
      config,
    });
  }

  static createEvent(provider, item, config = {}) {
    return RustCrateBridge.invoke("or-tools-productivity", "create_event", {
      provider,
      item,
      config,
    });
  }

  static listIssues(provider, query = {}, config = {}) {
    return RustCrateBridge.invoke("or-tools-productivity", "list_issues", {
      provider,
      query,
      config,
    });
  }

  static createIssue(provider, item, config = {}) {
    return RustCrateBridge.invoke("or-tools-productivity", "create_issue", {
      provider,
      item,
      config,
    });
  }

  static searchPages(provider, query = {}, config = {}) {
    return RustCrateBridge.invoke("or-tools-productivity", "search_pages", {
      provider,
      query,
      config,
    });
  }

  static createPage(provider, item, config = {}) {
    return RustCrateBridge.invoke("or-tools-productivity", "create_page", {
      provider,
      item,
      config,
    });
  }

  static postMessage(provider, channel, text, config = {}) {
    return RustCrateBridge.invoke("or-tools-productivity", "post_message", {
      provider,
      channel,
      text,
      config,
    });
  }

  static searchMessages(provider, query = {}, config = {}) {
    return RustCrateBridge.invoke("or-tools-productivity", "search_messages", {
      provider,
      query,
      config,
    });
  }
}

File: bindings/dart/lib/src/tools.dart
import "bridge.dart";
import "types.dart";

final class SearchTools {
  static Object? search(String provider, JsonObject query, {JsonObject config = const <String, Object?>{}}) {
    return RustCrateBridge.invoke("or-tools-search", "search", <String, Object?>{
      "provider": provider,
      "query": query,
      "config": config,
    });
  }
}

final class WebTools {
  static Object? fetch(String provider, JsonObject request, {JsonObject config = const <String, Object?>{}}) {
    return RustCrateBridge.invoke("or-tools-web", "fetch", <String, Object?>{
      "provider": provider,
      "request": request,
      "config": config,
    });
  }

  static Object? scrape(String provider, String url, {JsonObject config = const <String, Object?>{}}) {
    return RustCrateBridge.invoke("or-tools-web", "scrape", <String, Object?>{
      "provider": provider,
      "url": url,
      "config": config,
    });
  }
}

final class VectorTools {
  static Object? ensureCollection(String provider, JsonObject data, {JsonObject config = const <String, Object?>{}}) {
    return RustCrateBridge.invoke("or-tools-vector", "ensure_collection", <String, Object?>{
      "provider": provider,
      "data": data,
      "config": config,
    });
  }

  static Object? upsert(String provider, JsonObject data, {JsonObject config = const <String, Object?>{}}) {
    return RustCrateBridge.invoke("or-tools-vector", "upsert", <String, Object?>{
      "provider": provider,
      "data": data,
      "config": config,
    });
  }

  static Object? delete(String provider, JsonObject data, {JsonObject config = const <String, Object?>{}}) {
    return RustCrateBridge.invoke("or-tools-vector", "delete", <String, Object?>{
      "provider": provider,
      "data": data,
      "config": config,
    });
  }

  static Object? query(String provider, JsonObject data, {JsonObject config = const <String, Object?>{}}) {
    return RustCrateBridge.invoke("or-tools-vector", "query", <String, Object?>{
      "provider": provider,
      "data": data,
      "config": config,
    });
  }
}

final class LoaderTools {
  static Object? load(JsonObject request) {
    return RustCrateBridge.invoke("or-tools-loaders", "load", <String, Object?>{
      "request": request,
    });
  }
}

final class ExecTools {
  static Object? execute(
    JsonObject request, {
    List<String> providers = const <String>["python", "shell"],
    JsonObject config = const <String, Object?>{},
  }) {
    return RustCrateBridge.invoke("or-tools-exec", "execute", <String, Object?>{
      "request": request,
      "providers": providers,
      "config": config,
    });
  }
}

final class FileTools {
  static Object? read(String path, {String provider = "local", JsonObject config = const <String, Object?>{}}) {
    return RustCrateBridge.invoke("or-tools-file", "read", <String, Object?>{
      "provider": provider,
      "path": path,
      "config": config,
    });
  }

  static Object? write(String path, String content, {String provider = "local", JsonObject config = const <String, Object?>{}}) {
    return RustCrateBridge.invoke("or-tools-file", "write", <String, Object?>{
      "provider": provider,
      "path": path,
      "content": content,
      "config": config,
    });
  }

  static Object? list(String path, {String provider = "local", JsonObject config = const <String, Object?>{}}) {
    return RustCrateBridge.invoke("or-tools-file", "list", <String, Object?>{
      "provider": provider,
      "path": path,
      "config": config,
    });
  }

  static Object? delete(String path, {String provider = "local", JsonObject config = const <String, Object?>{}}) {
    return RustCrateBridge.invoke("or-tools-file", "delete", <String, Object?>{
      "provider": provider,
      "path": path,
      "config": config,
    });
  }

  static Object? fetch(String provider, JsonObject query, {JsonObject config = const <String, Object?>{}}) {
    return RustCrateBridge.invoke("or-tools-file", "fetch", <String, Object?>{
      "provider": provider,
      "query": query,
      "config": config,
    });
  }
}

final class CommsTools {
  static Object? send(
    String provider,
    String to,
    String body, {
    String? from,
    JsonObject config = const <String, Object?>{},
  }) {
    return RustCrateBridge.invoke("or-tools-comms", "send", <String, Object?>{
      "provider": provider,
      "to": to,
      "body": body,
      "from": from,
      "config": config,
    });
  }
}

final class ProductivityTools {
  static Object? listEmails(String provider, {JsonObject query = const <String, Object?>{}, JsonObject config = const <String, Object?>{}}) {
    return RustCrateBridge.invoke("or-tools-productivity", "list_emails", <String, Object?>{
      "provider": provider,
      "query": query,
      "config": config,
    });
  }

  static Object? sendEmail(String provider, JsonObject item, {JsonObject config = const <String, Object?>{}}) {
    return RustCrateBridge.invoke("or-tools-productivity", "send_email", <String, Object?>{
      "provider": provider,
      "item": item,
      "config": config,
    });
  }

  static Object? listEvents(String provider, {JsonObject query = const <String, Object?>{}, JsonObject config = const <String, Object?>{}}) {
    return RustCrateBridge.invoke("or-tools-productivity", "list_events", <String, Object?>{
      "provider": provider,
      "query": query,
      "config": config,
    });
  }

  static Object? createEvent(String provider, JsonObject item, {JsonObject config = const <String, Object?>{}}) {
    return RustCrateBridge.invoke("or-tools-productivity", "create_event", <String, Object?>{
      "provider": provider,
      "item": item,
      "config": config,
    });
  }

  static Object? listIssues(String provider, {JsonObject query = const <String, Object?>{}, JsonObject config = const <String, Object?>{}}) {
    return RustCrateBridge.invoke("or-tools-productivity", "list_issues", <String, Object?>{
      "provider": provider,
      "query": query,
      "config": config,
    });
  }

  static Object? createIssue(String provider, JsonObject item, {JsonObject config = const <String, Object?>{}}) {
    return RustCrateBridge.invoke("or-tools-productivity", "create_issue", <String, Object?>{
      "provider": provider,
      "item": item,
      "config": config,
    });
  }

  static Object? searchPages(String provider, {JsonObject query = const <String, Object?>{}, JsonObject config = const <String, Object?>{}}) {
    return RustCrateBridge.invoke("or-tools-productivity", "search_pages", <String, Object?>{
      "provider": provider,
      "query": query,
      "config": config,
    });
  }

  static Object? createPage(String provider, JsonObject item, {JsonObject config = const <String, Object?>{}}) {
    return RustCrateBridge.invoke("or-tools-productivity", "create_page", <String, Object?>{
      "provider": provider,
      "item": item,
      "config": config,
    });
  }

  static Object? postMessage(String provider, String channel, String text, {JsonObject config = const <String, Object?>{}}) {
    return RustCrateBridge.invoke("or-tools-productivity", "post_message", <String, Object?>{
      "provider": provider,
      "channel": channel,
      "text": text,
      "config": config,
    });
  }

  static Object? searchMessages(String provider, {JsonObject query = const <String, Object?>{}, JsonObject config = const <String, Object?>{}}) {
    return RustCrateBridge.invoke("or-tools-productivity", "search_messages", <String, Object?>{
      "provider": provider,
      "query": query,
      "config": config,
    });
  }
}
